File: jet_box_size.fs
let packageBoxing package boxes =
    let vol x = List.fold ( * ) 1 x in
    let pack = List.sort package in
    let smaller p b = vol p < vol b in
    let fits b = List.zip pack b |>
        List.forall (fun (x,y) -> x <= y)
    in
    let fold_f (r, i, ob) b =
        let box = List.sort b in
        let isSmaller = ob
            |> Option.map (smaller box)
            |> Option.getOrElse true
        in
        if fits box && isSmaller
        then (i, i+1, Some box)
        else (r, i+1, ob)
    in
    let (r,_,_) = List.fold fold_f ((-1), 0, None) boxes in
    r

File: jet_box_size.hs
import Data.List
import Data.Maybe
import Data.Function
import Data.Ord

packageBoxing pkg boxes
  = fromMaybe (-1)
  . fmap snd
  . listToMaybe
  . filter (fits (sort pkg) . sort . fst)
  . sortBy (comparing (product . fst))
  $ zip boxes [0..]
 where
  fits      pkg box = checkVol pkg box && checkSize pkg box
  checkVol  pkg box = product pkg <= product box
  checkSize pkg box = all id (zipWith (<=) pkg box)
